File: csharp/PrototypePattern/DeepStudent.cs

namespace PrototypePattern
{
    /// <summary>
    /// 深拷贝
    /// </summary>
    public class DeepStudent : AbstractStudent
    {
        public override AbstractStudent Clone()
        {
            DeepStudent student = new DeepStudent();
            student.Name = base.Name;
            student.Age = base.Age;
            student.Course = new Course();
            student.Course.Name = base.Course.Name;

            return student;
        }   
    }
}

File: c++/PrototypePattern/DeepStudent.cpp
#include "stdafx.h"
#include "DeepStudent.h"

DeepStudent::DeepStudent()
{

}

DeepStudent::DeepStudent(const DeepStudent &student)
{
	this->name = student.name;
	this->age = student.age;
	this->course = new Course();
	this->course->name = student.course->name;
}

Student *DeepStudent::Clone()
{
	return new DeepStudent(*this);
}